architecture/fxa.md

```diff
@@ -0,0 +1,252 @@
+# Integration with Firefox Accounts #
+
+Integration with Firefox Accounts (FxA) uses [OAuth](https://oauth.net/); Lockbox applications (Web Extension and mobile) are OAuth public clients and FxA is the authorization service.  It also functions as the user profile service where that information is needed.
+
+## OAuth Protocol Highlights ##
+
+**NOTE:** The OAuth details herein proscribe details and assumptions for Lockbox applications, defining specifics to the options that OAuth allows.
+
+FxA supports [OAuth 2.0][RFC6749] Authorization Code Grant flow, with some additions:
+
+* [PKCE][PKCE] proof keys, using the `S256` challenge mode
+* Application-scoped encryption keys derived from the user's credentials
+
+The complete interactive sign in flow for an FxA OAuth client is:
+
+1. Send an HTTPS `GET` request via a user agent to the authorization endpoint with query parameters
+2. The user agent interacts with the user to complete the authorization request
+3. Upon success, the FxA authorization endpoint redirects to the client's `redirect_uri`, including results as query parameters
+4. The client requests an exchange of the authorization code for access (and refresh) tokens from the token endpoint using an HTTPS `POST` request
+5. The FxA token endpoint validates the request and (upon success) provides the requested tokens
+6. If user information is needed, the client sends an HTTPS `GET` request to the FxA userinfo endpoint (providing the access token as authorization)
+7. The FxA userinfo endpoint validates the authorization and (upon success) provides the user's profile information
+
+The hosted FxA deployments support [endpoint discovery](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction#Endpoint_Discovery) as well as [documented details](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Firefox_Accounts/Introduction#Firefox_Accounts_deployments).
+
+### Authentication and Authorization ###
+
+The authorization portion starts by sending an HTTPS `GET` request to the FxA authorization endpoint (`/v1/authorization`) via a web user agent (e.g., web browser).  The request includes following as query parameters:
```

Comment:
Please look up these URLs via endpoint discovery mentioned above, rather than hard-coding them in your application.  You should be able to find all the URLs you need given only the root issuer URL "https://accounts.firefox.com", and if there's a URL you *can't* look up from there, I'll consider it a bug.

Reply:
The call outs to the endpoints was only meant tie things together, not that they will be hardcoded.  In the next iteration of this document I'll make it clearer to use endpoint discovery.

As far as I can tell, everything we need to talk to FxA-wise is discoverable.